stream-chat-android-state/src/main/java/io/getstream/chat/android/offline/plugin/logic/channel/internal/TypingEventPruner.kt

```diff
@@ -0,0 +1,210 @@
+    ) -> Unit,
+) {
+
+    /**
+     * A mutable map of typing events.
+     *
+     * Each event is capable of timed death using [TimedTypingStartEvent.removeTypingEvent]
+     */
+    private val typingEvents = mutableMapOf<String, TimedTypingStartEvent>()
+
+    /**
+     * Called when a typing event needs to be processed.
+     *
+     * Send in a null value of [typingStartEvent] to signal that
+     * a typing event should be removed.
+     *
+     * Please take care not to send in a typing event if the currently
+     * logged in user is typing.
+     *
+     * @param userId The ID of the user tied to the typing event.
+     * @param typingStartEvent The typing event that is being processed.
+     */
+    fun processEvent(
+        userId: String,
+        typingStartEvent: TypingStartEvent?,
+    ) {
+        when (typingStartEvent) {
+            null -> removeTypingEvent(userId)
+            else -> addTypingEvent(
+                userId = userId,
+                typingStartEvent = typingStartEvent
+            )
+        }
+    }
+
+    /**
+     * Adds a typing event to [typingEvents].
+     *
+     * It will cancel the timed death of an typing event with the same key (userId)
+     * if such exists.
+     *
+     *
+     * @param userId The ID of the user tied to the typing event.
+     * @param typingStartEvent The typing event will be added to [typingEvents].
+     */
+    private fun addTypingEvent(userId: String, typingStartEvent: TypingStartEvent) {
+        // Create a new self stopping event
+        val timedTypingStartEvent = TimedTypingStartEvent(
+            coroutineScope = coroutineScope,
+            typingStartEvent = typingStartEvent,
+            userId = userId,
+            delayTimeMs = DEFAULT_DELAY_TIME_MS,
+            removeTypingEvent = {
+                removeTypingEvent(it)
+            }
+        )
+
+        // Cancel the self stopping event you are replacing if one exists
+        typingEvents.getOrDefault(userId, null)?.cancelJob()
+
+        // Replace the old self stopping event and call
+        // the updated typing events listener
+        typingEvents[userId] = timedTypingStartEvent
+        onUpdated(getRawTyping(), getTypingEvent())
+    }
+
+    /**
+     * Removes a typing event with the given key (userId)
+     * if such exists in [typingEvents].
+     *
+     * @param userId The ID of the user tied to the typing event.
+     */
+    private fun removeTypingEvent(userId: String) {
+        typingEvents.getOrDefault(userId, null)?.cancelJob()
+
+        typingEvents.remove(userId)
+        onUpdated(getRawTyping(), getTypingEvent())
+    }
+
+    /**
+     * Produces an up to date raw typing map.
+     *
+     * Used to set a value to
+     * [io.getstream.chat.android.offline.plugin.state.channel.internal.ChannelMutableState.rawTyping].
+     */
+    private fun getRawTyping(): Map<String, TypingStartEvent> = typingEvents.mapValues { it.value.typingStartEvent }
```

Comment:
```suggestion
    private fun getRawTypingEvents(): Map<String, TypingStartEvent> = typingEvents.mapValues { it.value.typingStartEvent }
```

Reply:
Decided to keep due to naming consistency with `ChannelMutableState.rawTyping`.